boards: phytec: Add support for phyCORE-RT1170

dnltz opened the pull request:
This is another attempt to add support for the phyBOARD-Atlas development kit featuring the phyCORE-RT1170 SOM.

While #86756 had both Cortex-M CPUs included, this PR only adds the M7 and partly prepares this board for adding the M4 later.  We first need to figure out how to properly integrate the M4 in our development kit as well as testing infrastructure.

Depends on the following PRs being merged first:
- [x] https://github.com/zephyrproject-rtos/zephyr/pull/95629

FPlohl committed:
dts: arm: phytec: add support for phyCORE-RT1170 SOM

Add phyCORE-1170 SOM dtsi for streamlined inclusion on carrier
boards and uses a dual-core NXP i.MX RT1170 SoC as a basis.

Signed-off-by: Florijan Plohl <florijan.plohl@norik.com>
Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz committed:
boards: phytec: add support for phyBOARD-Atlas i.MX RT1170

Add initial support for PHYTEC phyBOARD-Atlas i.MX RT1170 based on the
PHYTEC phyCORE-i.MX RT1170 SOM. SOM uses dual-core NXP i.MX RT1170 SoC
as a basis, with Cortex-M7 core running at 1 GHz and Cortex-M4 core at
400 MHz.

Supported features:
 * Accelerometer
 * Audio codec
 * CAN
 * DAC
 * EEPROM
 * Ethernet
 * External display
 * I2C
 * RS-232
 * SD-Card
 * SPI
 * UART
 * USB

Signed-off-by: John Ma <jma@phytec.com>
Signed-off-by: Florijan Plohl <florijan.plohl@norik.com>
Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz commented:
@brycewilkins I’ve finally wrapped up the RT1170 pull request. Your review would be greatly appreciated.

nordicjm commented on dts/arm/phytec/phycore_rt1170_common.dtsi:
```diff
@@ -0,0 +1,135 @@
+			/* Note slot 0 has one additional sector,
+			 * this is intended for use with the swap move algorithm
+			 */
+			slot0_partition: partition@20000 {
```
missing newline gaps between nodes

nordicjm commented on boards/phytec/phyboard_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+if(CONFIG_NXP_IMXRT_BOOT_HEADER)
+  zephyr_library()
+  set(RT1170_BOARD_NAME "phycore_rt1170")
+  if(CONFIG_BOOT_FLEXSPI_NOR)
```
add spacing between these blocks so it's not all squashed

nordicjm commented on boards/phytec/phyboard_atlas/board.cmake:
```diff
@@ -0,0 +1,29 @@
+ board_runner_args(linkserver "--core=cm4")
+endif()
+
+if(NOT SYSBUILD)
+  board_runner_args(jlink "--no-reset")
+  board_runner_args(linkserver "--no-reset")
+endif()
```
this should not be doing sysbuild exceptions, what is this meant to do?

dnltz replied:
I honestly just synced that from NXP's board.cmake file: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/mimxrt1170_evk/board.cmake#L30

I would like to keep our board similar to NXP's.

nordicjm replied:
@EmilioCBen can you explain this please, there should be no knowledge or care if sysbuild is being used in a board.cmake file, and this needs to go from your board

dnltz replied:
@nordicjm I just reverted this change from this board.cmake file.

EmilioCBen replied:
@nordicjm The if statement originally comes from this PR files[87423](https://github.com/zephyrproject-rtos/zephyr/pull/87423/files)

The use case was that a user wanted to debug something that was flashed to ram but did not want to remove the image that was inside the flash. I added the sysbuild check when anticipating that sysbuild usually requires a flash. I am unable to test this myself on the 1170 as sysbuild does not seem to work or has not prior been enabled.

I have a removal PR for this check right here. [96038](https://github.com/zephyrproject-rtos/zephyr/pull/96038)
..Sorry about the delay response, I thought I might have been able to enable the sysbuild as well..

nordicjm commented on boards/phytec/phyboard_atlas/phyboard_atlas-pinctrl.dtsi:
```diff
@@ -0,0 +1,324 @@
+			bias-pull-down;
+			slew-rate = "fast";
+		};
+		group1 {
```
as above, fix in whole PR

nordicjm commented on boards/phytec/phyboard_atlas/phyboard_atlas_mimxrt1176_cm7.dts:
```diff
@@ -0,0 +1,152 @@
+		#gpio-cells = <2>;
+		gpio-map-mask = <0xffffffff 0xffffffc0>;
+		gpio-map-pass-thru = <0 0x3f>;
+		gpio-map =	<9 0 &gpio11 15 0>,	/* Pin 9, RESETB */
```
gap after `=` seems large, change to 1 space

nordicjm commented on boards/phytec/phyboard_atlas/xip/phycore_rt1170_flexspi_nor_config.h:
```diff
@@ -0,0 +1,381 @@
+	FLEXSPI_LUT_NUM_PADS1(pad1) | FLEXSPI_LUT_OPCODE1(cmd1))
+
+/*! @brief Definitions for FlexSPI Serial Clock Frequency */
+typedef enum _FlexSpiSerialClockFreq {
```
this file does not comply with coding guidelines, snake case variables, other files too

nordicjm replied:
it's not that I have blocked it (I can dismiss review if the other issues are fixed), I get that we don't have a CI check for it (and it's likely impossible to have one) but the files should never have been added in the first place because they do not comply with the requirements https://docs.zephyrproject.org/latest/contribute/style/code.html so the issue has been brought up @kartben

dnltz replied:
@ZhaoxiangJin thanks for fixing this issue!

dnltz replied:
@nordicjm same here. Update both files in `xip/` with those changed from #96120 . Should also be resolved.

dnltz committed:
tests: dac: Add phyBOARD-Atlas

phyBOARD-Atlas is based on NXP's RT1176 MCU. Add this
board nex to the NXP defines.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -62,7 +62,7 @@ struct mc_ksz8081_config {
 #define KSZ8081_LINK_STATE_VALID BIT(2)
 
-#define USING_INTERRUPT_GPIO							\
+#define USING_INTERRUPT_GPIO(config)						\
```
Yes, I thought it's gonna fix the "config is unused" error, since config is actually used then. Will rebase as soon as #95629 got merged. Thanks for referencing this PR.

dnltz committed:
samples: net: prometheus: Exclude phyBOARD-Atlas

This sample doesn't compile for the phyBOARD-Atlas. Exlude this board
until it got fixed.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

maass-hamburg commented on boards/phytec/phyboard_atlas/Kconfig.defconfig:
```diff
@@ -0,0 +1,44 @@
+if DISK_DRIVERS
+
+config IMX_USDHC_DAT3_PWR_TOGGLE
+	default y
```
by default this is already enabled

dnltz replied:
ah, thanks. Removed the entire  `DISK_DRIVERS` section.

maass-hamburg commented on boards/phytec/phyboard_atlas/Kconfig.defconfig:
```diff
@@ -0,0 +1,44 @@
+
+endif # DISK_DRIVERS
+
+if NETWORKING
+
+config NET_L2_ETHERNET
+	default y if CPU_CORTEX_M7 # No cache memory support is required for driver
+
+endif # NETWORKING
```
use a configdefault here

maass-hamburg commented on boards/phytec/phyboard_atlas/Kconfig.defconfig:
```diff
@@ -0,0 +1,44 @@
+if BOARD_PHYBOARD_ATLAS
+
+# Use External Memory Configuration Data (XMCD) by default when booting primary core (M7)
+config EXTERNAL_MEM_CONFIG_DATA
+	default y if CPU_CORTEX_M7
+
+config NXP_IMX_EXTERNAL_SDRAM
+	default y if CPU_CORTEX_M7
```
might be better to do the CPU_CORTEX_M7 dependencies around the configs
```
if CPU_CORTEX_M7
configdefault EXTERNAL_MEM_CONFIG_DATA
	default y
....
endif
```

maass-hamburg commented on boards/phytec/phyboard_atlas/Kconfig.defconfig:
```diff
@@ -0,0 +1,41 @@
+config NXP_IMX_EXTERNAL_SDRAM
+	default y
+
+if NETWORKING
+
+config NET_L2_ETHERNET
+	default y # No cache memory support is required for driver
+
+endif # NETWORKING
```
@dnltz when there are multiple entries that need `if CPU_CORTEX_M7` use the second, if there is only one use, then the first.

dnltz commented on boards/phytec/phyboard_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if(CONFIG_NXP_IMXRT_BOOT_HEADER)
+  zephyr_library()
+  set(RT1170_BOARD_NAME "phycore_rt1170")
```
I would like to keep it as is it because it's identical to NXP's RT1170 board right now. I don't know if we run in some special issues when moving it outside.

nordicjm commented on boards/phytec/phyboard_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+  set(RT1170_BOARD_NAME "phycore_rt1170")
+
+  if(CONFIG_BOOT_FLEXSPI_NOR)
+    zephyr_compile_definitions(XIP_BOOT_HEADER_ENABLE=1)
+    zephyr_compile_definitions(BOARD_FLASH_SIZE=CONFIG_FLASH_SIZE*1024)
```
not fixed

ZhaoxiangJin replied:
I'd like to fix this for all NXP platforms, but I may not be able to do it in time, as I'm taking a holiday recently and will be back to work on October 9.😄

nordicjm replied:
@decsny please reassign to someone

dnltz replied:
As soon as I change those definitions to `zephyr_library_compile_definitions`, I get following build error:
```
In file included from /.../zephyr/soc/nxp/imxrt/imxrt11xx/soc.c:23:
/.../modules/hal/nxp/mcux/mcux-sdk-ng/devices/RT/RT1170/MIMXRT1176/xip/fsl_flexspi_nor_boot.h:14:10: fatal error: board.h: No such file or directory
   14 | #include "board.h"
      |  
```
So, these defines are not passed to the HAL anymore.

Changes made in #96902 result in multiple NXP boards not booting anymore.

I really think this is a bigger task and nothing I can fix here without touching other NXP boards.

hakehuang replied:
> also, why do these flags need passing to everything? Maybe should be zephyr_library_compile_definitions and below too

passing flags global make the setting consist for all linked modules, and I do not see any problem and side effects with this. but limited to modules, this could be introduce potential miss-use, or make core ambiguous to read. please guide me what is the real benefit to put defines in module only. Thanks

nordicjm replied:
this is still not addressed

nordicjm commented on boards/phytec/phyboard_atlas/board.cmake:
```diff
@@ -0,0 +1,27 @@
+#
+
+if(CONFIG_SOC_MIMXRT1176_CM7 OR CONFIG_SECOND_CORE_MCUX)
+ board_runner_args(linkserver "--no-reset")
```
2 space indent for cmake

dnltz replied:
Fixed.

nordicjm commented on boards/phytec/phyboard_atlas/xip/phycore_rt1170_flexspi_nor_config.h:
```diff
@@ -0,0 +1,375 @@
+
+/*! @brief Definitions for FlexSPI Serial Clock Frequency */
+typedef enum flexspi_serial_clock_freq {
+	flexspi_serial_clk_30mhz = 1,
```
enum values are uppercase https://kernel.org/doc/html/latest/process/coding-style.html#macros-enums-and-rtl

dnltz replied:
This file got updated in #96120 and is used by 18 other NXP boards with the same style-guide issue.

ZhaoxiangJin replied:
Hello @dnltz, sorry for that, at the time, the focus was solely on converting camel case to snake case. I will do a update again.

pefech reviewed (dismissed):
Looks good overall to me now

dnltz commented:
@maass-hamburg can you please check this PR again?

maass-hamburg reviewed (dismissed):
seems fine, but can't say much about the nxp specific stuff

dnltz commented:
> @dnltz can you please rebase this PR? A fix for @nordicjm's request was submitted here: #100252 This way, we can move fwd with this PR.

@iuliana-prodan Rebased and fixed the merge conflict. Also fixed some dts style guide issues.

nordicjm commented:
> @nordicjm NXP already opened #100252 to fix this issue. I really don't understand why you're blocking this pull-request while almost all NXP boards already use this definition. It's obviously not a critical issue and will be fixed soon.

I think they asked you to update your PR to follow that, which you have not done. Bearing in mind I pointed this out in this PR 3 months ago, bearing in mind it was blocked from being added in a posix update and bearing in mind other wrong usage of it has been cleaned up in the tree as https://github.com/zephyrproject-rtos/zephyr/issues/98647 then it is fine to ask you/nxp to do the same, that it has taken over 3 months and is still not done is nothing to do with me. Just because something exists that is bad, does not mean further prolification  of it should continue

nordicjm commented on boards/phytec/phyboard_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+  endif()
+
+  if(CONFIG_EXTERNAL_MEM_CONFIG_DATA)
+    zephyr_compile_definitions(XIP_BOOT_HEADER_XMCD_ENABLE=1)
```
this is still not fixed

nordicjm commented on boards/phytec/phyboard_atlas/phyboard_atlas-pinctrl.dtsi:
```diff
@@ -25,15 +25,15 @@
 		group0 {
 			pinmux = <&iomuxc_gpio_disp_b2_08_enet_rx_en>,
-				<&iomuxc_gpio_disp_b2_09_enet_rx_er>;
+				 <&iomuxc_gpio_disp_b2_09_enet_rx_er>;
```
no fixup commits, fix in original commit

dnltz committed:
tests: i2c: i2c_speed: Add phyBOARD-Atlas Files

Add an overlay and conf file for the phyBOARD-Atlas. Those
files are copies from NXP's RT1176 board, because the phyBOARD-Atlas
is using the same MCU.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>
